chirho/dynamical/internals/backends/diffeqdotjl.py

```diff
@@ -0,0 +1,558 @@
+    return np.concatenate([v.ravel() for v in vs])
+
+
+def _flatten_state_ao_params(state_ao_params: StateAndOrParams[Union[Tnsr, np.ndarray]]) -> Union[Tnsr, np.ndarray]:
+    if len(state_ao_params) == 0:
+        # TODO do17bdy1t address type specificity
+        return torch.tensor([], dtype=torch.float64)
+    var_order = get_var_order(state_ao_params)
+    return flat_cat(*[state_ao_params[v] for v in var_order])
+
+
+def _unflatten_state(
+        flat_state_ao_params: Tnsr,
+        shaped_state_ao_params: StateAndOrParams[Union[Tnsr, juliacall.VectorValue]],
+        to_traj: bool = False
+) -> StateAndOrParams[Union[Tnsr, np.ndarray]]:
+
+    var_order = get_var_order(shaped_state_ao_params)
+    state_ao_params = StateAndOrParams()
+    for v in var_order:
+        shaped = shaped_state_ao_params[v]
+        shape = shaped.shape
+        if to_traj:
+            # If this is a trajectory of states, the dimension following the original shape's state will be time,
+            #  and because we know the rest of the shape we can auto-detect its size with -1.
+            shape += (-1,)
+
+        sv = flat_state_ao_params[:shaped.numel()].reshape(shape)
+
+        state_ao_params[v] = sv
+
+        # Slice so that only the remaining elements are left.
+        flat_state_ao_params = flat_state_ao_params[shaped.numel():]
+
+    return state_ao_params
+
+
+def require_float64(state_ao_params: StateAndOrParams[Tnsr]):
+    # Forward diff through diffeqpy currently requires float64. # TODO do17bdy1t update when this is fixed.
+    for k, v in state_ao_params.items():
+        if v.dtype is not torch.float64:
+            raise ValueError(f"State variable {k} has dtype {v.dtype}, but must be float64.")
+
+
+def separate_state_and_params(dynamics: Dynamics[np.ndarray], initial_state_ao_params: StateAndOrParams[Tnsr], t0: Tnsr):
+    """
+    Non-explicitly (bad?), the initial_state must include parameters that inform dynamics. This is required
+     for this backend because the solve function passed to Julia must be a pure wrt to parameters that
+     one wants to differentiate with respect to.
+    """
+
+    # Copy so we can add time in without modifying the original, also convert elements to numpy arrays so that the
+    #  user's dynamics only have to handle numpy arrays, and not also torch tensors. This is fine, as the only way
+    #  we use the initial_dstate below is for its keys, which comprise the state variables.
+    initial_state_ao_params_np = {k: copy(v.detach().numpy()) for k, v in initial_state_ao_params.items()}
+    # TODO unify this time business with how torchdiffeq is doing it?
+    if 't' in initial_state_ao_params_np:
+        raise ValueError("Initial state cannot contain a time variable 't'. This is added on the backend.")
+    initial_state_ao_params_np['t'] = t0.detach().numpy()
+
+    # Run the dynamics on the converted initial state.
+    initial_dstate_np = dynamics(initial_state_ao_params_np)
```

Comment:
cc @SamWitty 

Do a check here to return a more informative message saying that dynamics for diffeqdotjl backend have to support numpy arrays